### PR TITLE
feat(encoder): implement token-aware batch semantic extraction

### LINKED ISSUE #81: Implement token-aware batch semantic extraction
## Summary

Replace the current fixed-size batch semantic extraction (5 entities per LLM call) with a **token-aware batching** strategy that groups entities until a token budget is reached. This is identified as a **P1 improvement** in the [vendor comparison analysis](docs/vendor-comparison.md#41-semantic-lifting).

### Current State
Our `SemanticExtractor` in `packages/encoder/src/semantic.ts` uses a fixed batch size of **5 entities per LLM call**, regardless of entity size. This means:
- Small entities waste context window capacity (5 tiny functions = few tokens, could fit many more)
- Large entities may exceed context limits (5 large classes could overflow)
- Suboptimal LLM cost efficiency for large repositories

### Vendor Reference
The vendor (`rpg_encoder/rpg_parsing/rpg_encoding.py`) uses token-aware batching with configurable budgets:

| Aspect | Vendor | Ours |
|--------|--------|------|
| **Batching** | Token-aware: `min_batch_tokens=10K`, `max_batch_tokens=50K` | Fixed batch size: 5 entities per call |
| **Iterations** | Multi-iteration: `max_parse_iters=10` | Single-pass extraction |
| **Parallelization** | Sequential within batch | Sequential |

The vendor groups entities until the cumulative token count reaches the `max_batch_tokens` threshold, ensuring each LLM call maximizes context window utilization without overflow.

### Impact
- **Large repositories**: Fewer LLM calls needed (better throughput)
- **Cost reduction**: More entities per call = fewer API calls = lower cost
- **Reliability**: Prevents context overflow for batches with large entities
- **Complementary**: Works alongside our existing `SemanticCache` (which the vendor lacks)

## Tasks

- [ ] Implement token counting utility
  - Estimate token count for source code + prompt template per entity
  - Support configurable tokenizer (tiktoken-compatible or character-based approximation)
- [ ] Refactor `SemanticExtractor` batching strategy
  - Replace fixed batch size with token budget parameters (`minBatchTokens`, `maxBatchTokens`)
  - Group entities greedily until `maxBatchTokens` is reached
  - Ensure at least one entity per batch (handle single-entity overflow gracefully)
- [ ] Add configurable batch parameters to `SemanticOptions`
  - `minBatchTokens?: number` (default: 10,000)
  - `maxBatchTokens?: number` (default: 50,000)
  - Maintain backward compatibility (fixed batch size as fallback)
- [ ] Implement multi-iteration extraction (optional, P2 follow-up)
  - Retry failed/incomplete extractions up to `maxParseIterations`
  - Vendor uses `max_parse_iters=10` for iterative refinement
- [ ] Update CLI `encode` command to expose batch parameters
  - `--min-batch-tokens` and `--max-batch-tokens` flags
- [ ] Write tests
  - Token counting accuracy
  - Batch grouping correctness (small entities grouped, large entities isolated)
  - Edge cases: single entity exceeding budget, empty entity list
  - Integration test with real LLM call verifying batching behavior

## References

- Vendor source: `vendor/RPG-ZeroRepo/zerorepo/rpg_encoder/rpg_parsing/rpg_encoding.py` (796 lines)
- Vendor comparison: [`docs/vendor-comparison.md` §4.1](docs/vendor-comparison.md)
- Our current implementation: `packages/encoder/src/semantic.ts` (`SemanticExtractor`)
- Semantic options: `packages/encoder/src/semantic.ts` (`SemanticOptions`)

### PR DESCRIPTION
**Pull request by amondnet**
## Summary

Implement token-aware batch semantic extraction replacing the fixed batch size of 5 entities per LLM call with a configurable token budget strategy.

- Add `estimateTokenCount`, `estimateEntityTokens`, `estimateBatchTokens` utilities for token estimation
- Refactor `SemanticExtractor.extractBatch()` to use greedy token-aware batching with `minBatchTokens` (default: 10K) and `maxBatchTokens` (default: 50K) thresholds
- Add `maxParseIterations` option for multi-iteration LLM extraction retry
- Expose `--min-batch-tokens` and `--max-batch-tokens` CLI flags in `encode` and `evolve` commands
- Validate CLI batch token arguments against NaN and negative values

## Verification Checklist

- [x] Token-aware batching replaces fixed batch size of 5
- [x] Configurable minBatchTokens (default 10000) and maxBatchTokens (default 50000)
- [x] At least one entity per batch even if it exceeds budget
- [x] Backward compatible: all 684 existing tests pass
- [x] CLI exposes --min-batch-tokens and --max-batch-tokens flags
- [x] TypeScript typecheck passes
- [x] Lint passes
- [x] 71 new tests covering all functionality

## Test plan

- [x] Token counting accuracy tests (17 tests)
- [x] Batch grouping correctness with deterministic boundary assertions (12 tests)
- [x] SemanticOptions batch parameter acceptance (22 tests)
- [x] Multi-iteration extraction retry logic (12 tests)
- [x] CLI flag integration (8 tests)

Closes #81

### CODE REVIEW

File: packages/encoder/src/semantic.ts [hunk 1]
```diff
@@ -137,22 +157,100 @@ export class SemanticExtractor {
...
+    if (batches.length > 1) {
+      const lastBatchTokens = batchTokenCounts[batchTokenCounts.length - 1]!
+
+      if (lastBatchTokens < minBatchTokens) {
+        const lastBatch = batches[batches.length - 1]!
+        const previousBatch = batches[batches.length - 2]!
+        previousBatch.push(...lastBatch)
+        batches.pop()
+      }
+    }
```
[the remark this answers is not in the document]
Comment (amondnet) [reply packages/encoder/src/semantic.ts]: ✅ Applied in [91b2d94](https://github.com/pleaseai/RPG/commit/91b2d94)
- File: `packages/encoder/src/semantic.ts`
- Change: Added `prevBatchTokens + lastBatchTokens <= maxBatchTokens` guard before merging, and now also pop from `batchTokenCounts` to keep both arrays in sync.

Thanks for catching both issues, @gemini-code-assist!

File: tests/semantic-retry.test.ts [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+        filePath: 'src/services/user.ts',
+      }
+
+      const feature = await extractor.extract(input)
+
+      expect(feature.description).toContain('define')
+      expect(feature.keywords).toContain('class')
+    })
+
+    it('preserves warnings array across multiple extractions', async () => {
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/semantic-retry.test.ts]: ✅ Applied in [91b2d94](https://github.com/pleaseai/RPG/commit/91b2d94)
- File: `tests/semantic-retry.test.ts`
- Change: Added a second `extract(input2)` call and asserted `warningsAfterSecond > warningsAfterFirst` to actually verify accumulation across multiple extractions.

Thanks, @cubic-dev-ai!

File: tests/token-counter.test.ts [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+          filePath: 'src/c.ts',
+          sourceCode: 'function process(data) { return data.map(x => x * 2); }',
+          documentation: 'Process data array',
+        },
+      ]
+      const count = estimateBatchTokens(entities)
+
+      // Entity 1: 0 (no source) + 200 = 200
+      // Entity 2: ceil(16/4) + 200 = 4 + 200 = 204
+      // Entity 3: ceil(53/4) + ceil(20/4) + 200 = 14 + 5 + 200 = 219
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/token-counter.test.ts]: ✅ Applied in [91b2d94](https://github.com/pleaseai/RPG/commit/91b2d94)
- File: `tests/token-counter.test.ts`
- Change: Corrected Entity 3 comment: source is 55 chars (`ceil(55/4)=14`) and doc is 18 chars (`ceil(18/4)=5`).

Thanks, @cubic-dev-ai!
